### DISCUSSION TITLE
How do I do nested TemplateBinding?

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#9880] [Q&A] [answered]
Following the various examples, I have a simple template binding object derived from `ContentControl`.
   
    <Styles ...>
      <Style Selector="controls|GraphGrid">
        <Setter Property="Template">
          <ControlTemplate>
            <ContentPresenter Content="{TemplateBinding Content}"/>
          </ControlTemplate>
        </Setter>
      </Style>

If I instantiate like this, then the Label is displayed:
    
    <GraphGrid>
        <Button>Click Me</Button>
    </GraphGrid>`

But if I try putting the templatebinding in a nested item then it doesn't show:
    
    <Styles ...>
      <Style Selector="controls|GraphGrid">
        <Setter Property="Template">
          <ControlTemplate>
            <ContentPresenter>
              <ContentPresenter.Content>
                <Panel>
                  <ContentPresenter Content="{TemplateBinding Content}" />
                </Panel>
              </ContentPresenter.Content>
            </ContentPresenter>
          </ControlTemplate>
        </Setter>
      </Style>

I want the Content passed in to the outer thing, ie `<Button>Click Me</Button>`, to show up in the inner thing.

The DevTools viewer shows the Button in the Logical Tree, as a child of GraphGrid, which is correct, but the Button's bounds are all zero.
![image](https://user-images.githubusercontent.com/672429/210713580-7bcaa969-cda1-4f13-8210-2efa4ccbcc94.png)

However the Visual Tree shows the inner ContentPresenter is empty, which is weird.  How did the button become a visual child of the GraphGrid if the ContentPresenter is empty?
![image](https://user-images.githubusercontent.com/672429/210713234-b4e202dc-b57c-4dd5-a07c-a61460e49539.png)

Maybe I need some syntax magic similar to `{Binding Content, RelativeSource={RelativeSource TemplatedParent}}`.

thanks

**Comment by timunie** [comment, marked as answer]
`ContentPresenster` is a special one. The only puspose is to work for presenting the user content. You should not nest them. My suggestion is: 

```xml
<Styles ...>
  <Style Selector="controls|GraphGrid">
    <Setter Property="Template">
      <ControlTemplate>
        <ContentControl>
          <ContentControl.Content>
            <Panel>
              <ContentPresenter Content="{TemplateBinding Content}" />
            </Panel>
          </ContentControl.Content>
        </ContentControl>
      </ControlTemplate>
    </Setter>
  </Style>
```

Also you can shorten it to this, if there is no other property set on the outer `ContentControl`: 

```xml
<Styles ...>
  <Style Selector="controls|GraphGrid">
    <Setter Property="Template">
      <ControlTemplate>
            <Panel>
              <ContentPresenter Content="{TemplateBinding Content}" />
            </Panel>
      </ControlTemplate>
    </Setter>
  </Style>
```

**Reply by sonicsmooth** [reply]
Thanks, will see if I can work that in.
